### PR TITLE
[TIR] An analysis pass to calculate workspace size for primfuncs

### PR DESCRIPTION
**Pull request by manupa-arm**
This commit introduces functionality to query the workspace sizeas required by a tir primfunc by looking at tir.allocates inside of it.
I needed to do a small fix for tvmscript parser and scope handler to make the test cases work.

### CODE REVIEW

File: include/tvm/tir/analysis.h [hunk 1]
```diff
@@ -178,6 +178,12 @@ Array<Array<BufferRegion>> GetBlockAccessRegion(const Block& block,
  */
 TVM_DLL size_t CalculateExprComplexity(const PrimExpr& expr);
 
+/*!
+ * \brief Calculate the workspace size in bytes needed by the TIR allocates inside the TIR PrimFunc
+ * \param func The TIR PrimFunc for which the workspace size to be calculated
+ */
+TVM_DLL int CalculateWorkspaceBytes(const PrimFunc& func);
```
Comment (jcf94) [inline include/tvm/tir/analysis.h:185]:
```suggestion
TVM_DLL size_t CalculateWorkspaceBytes(const PrimFunc& func);
```

File: src/tir/analysis/calculate_workspace.cc [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  void VisitStmt_(const AllocateNode* op) override;
+  int CalculateExtentsSize(const AllocateNode* op);
+  int current_size = 0;
+  int max_size = 0;
+};
+
+int WorkspaceCalculator::operator()(const PrimFunc& func) {
+  this->VisitStmt(func->body);
+  return this->max_size;
+}
```
Comment (jcf94) [inline src/tir/analysis/calculate_workspace.cc:47]: ditto. Update `int` to `size_t`

**Review by jcf94 — CHANGES_REQUESTED** [review]
Thanks! @manupa-arm Overall looks good.
Is the size guaranteed to be non-zero? I think its better to use `size_t` rather than the int.

File: include/tvm/tir/analysis.h (hunk 1 above)
Comment (manupa-arm) [reply include/tvm/tir/analysis.h]: Done.

File: src/tir/analysis/calculate_workspace.cc (hunk 2 above)
Comment (manupa-arm) [reply src/tir/analysis/calculate_workspace.cc]: Done.

**Comment by manupa-arm** [comment]
Thanks @jcf94 for the review -- I think I've addressed your comments.

**Review by jcf94 — APPROVED** [review]
LGTM. Thanks! @manupa-arm By the way, seems this will be used in your TVM AOT project?

**Comment by manupa-arm** [comment]
@jcf94 Yes, we are planning to use it there and we think it'd be a good utility for memory planning project as well.

**Pull request merged**